Handle the case when a Bignum in JRuby contains a small value

iconara opened the pull request:
This fixes #112.

In JRuby a Java extension can create a `Bignum` with any value, which our current implementation doesn't expect.

Ensure that the code handles byte arrays from Java's `BigInteger#toByteArray` that are less than 8 bytes – and make sure the top bytes are filled with either `0x00` or `0xFF` depending on the sign of the value.

I've included tests for this in `spec/msgpack_spec.rb`, but guarded so that they run only in JRuby, is this the right place for these kinds of platform-specific tests?

Can the same thing happen in MRI by the way? I assume there's a way to construct a `Bignum` with small values there too?

It would be possible to optimize the case when a `Bignum` has a small value and encode it as an `INT32`/`UINT32` or even smaller, but I don't know if we want to do that. `Bignum`s aren't supposed to contain numbers small enough to fit in an `INT32`, so optimizing that case might lead to confusing results, what do you think?

tagomoris commented:
We can create Bignum instance with 0 (small integer) in MRI too as I commented.
https://github.com/msgpack/msgpack-ruby/issues/112#issuecomment-217895259

So that we can make a new tests as common one among implementations.

tagomoris commented:
MessagePack int64 do NOT have any limitations for real value. It can store values which are in int32 range.
IMO, it's ok to pack Bignum with 32bit range value into int64. It consumes a bit more space on memory and disk, but it's not prohibited, and some implementations do same thins for float32/64 too.

iconara commented:
@tagomoris great that you knew a way to create small bignums in pure Ruby. I've rewritten the tests using `Bignum#coerce` and removed the Java guard.

I'm not sure I can fix the problem in MRI, but I can give it a go, unless you've already solved it?

tagomoris commented:
I tried the code below in MRI, and confirmed that there's no problem:
```
MBA:~ tagomoris$ ruby -v
ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-darwin15]
MBA:~ tagomoris$ pry
[1] pry(main)> require 'msgpack'
=> true
[2] pry(main)> MessagePack.pack((1 << 64).coerce(0)[0])
=> "\x00"
[3] pry(main)> MessagePack.unpack(MessagePack.pack((1 << 64).coerce(0)[0]))
=> 0
```

tagomoris commented:
Ah, MRI implementation packs that object into int8 (as small as possible).
It looks ok just to `unpack(pack(obj)) == obj`, isn't it?
Or, how about to add `if` to check produced binary with environments.

iconara commented:
I can rewrite the Java code to pack to pack to the smallest representation, and write tests that work for both implementations.

iconara commented:
I've updated the JRuby implementation to work like in MRI, and changed the tests to run the same tests in both JRuby and MRI.

tagomoris commented:
Looks very good to me.

tagomoris commented:
I released this fix as v0.7.6.
